src/OpenTelemetry/Logs/Pool/LogRecordSharedPool.cs

```diff
@@ -0,0 +1,180 @@
+
+        internal static LogRecordSharedPool Current = new(DefaultMaxPoolSize);
+
+        internal readonly int Capacity;
+        private readonly LogRecord?[] pool;
+        private long rentIndex;
+        private long returnIndex;
+
+        internal LogRecordSharedPool(int capacity)
+        {
+            this.Capacity = capacity;
+            this.pool = new LogRecord?[capacity];
+        }
+
+        internal int Count => (int)(Volatile.Read(ref this.returnIndex) - Volatile.Read(ref this.rentIndex));
+
+        public LogRecord Rent()
+        {
+            while (true)
+            {
+                var rentSnapshot = Volatile.Read(ref this.rentIndex);
+                var returnSnapshot = Volatile.Read(ref this.returnIndex);
+
+                if (rentSnapshot >= returnSnapshot)
+                {
+                    break; // buffer is empty
+                }
+
+                if (Interlocked.CompareExchange(ref this.rentIndex, rentSnapshot + 1, rentSnapshot) == rentSnapshot)
+                {
+                    var logRecord = Interlocked.Exchange(ref this.pool[rentSnapshot % this.Capacity], null);
+                    if (logRecord == null && !this.TryRentCoreRare(rentSnapshot, out logRecord))
+                    {
+                        continue;
+                    }
+
+                    logRecord.PoolReferenceCount = 1;
+                    return logRecord;
+                }
+            }
+
+            return new LogRecord()
+            {
+                PoolReferenceCount = 1,
+            };
+        }
+
+        public void Return(LogRecord logRecord)
+        {
+            if (Interlocked.Decrement(ref logRecord.PoolReferenceCount) != 0)
+            {
+                return;
+            }
+
+            Clear(logRecord);
+
+            while (true)
+            {
+                var rentSnapshot = Volatile.Read(ref this.rentIndex);
+                var returnSnapshot = Volatile.Read(ref this.returnIndex);
+
+                if (returnSnapshot - rentSnapshot >= this.Capacity)
+                {
+                    return; // buffer is full
+                }
+
+                if (Interlocked.CompareExchange(ref this.returnIndex, returnSnapshot + 1, returnSnapshot) == returnSnapshot)
+                {
+                    // If many threads are hammering rent/return it is possible
+                    // for two threads to write to the same index. In that case
+                    // only one of the logRecords will make it back into the
+                    // pool. Anything lost in the race will collected by the GC
+                    // and the pool will issue new instances as needed
+                    this.pool[returnSnapshot % this.Capacity] = logRecord;
+                    return;
+                }
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void TrackReference(LogRecord logRecord)
```

Comment:
Could this be static?

Reply:
Or maybe move it to `LogRecord.AddRef`?

Reply:
Looks good to me, I have a minor comment regarding naming https://github.com/open-telemetry/opentelemetry-dotnet/pull/3385#discussion_r903112020.